src/main.rs

```diff
@@ -1,3 +1,1827 @@
+            }
+        }
+
+        Self {
+            list,
+            index,
+            default_criteria,
+        }
+    }
+    fn criteria_from_entry(&self, entry: &AuditEntry) -> CriteriaSet {
+        if let Some(criteria_list) = entry.criteria.as_ref() {
+            self.criteria_from_list(criteria_list.iter().map(|s| &**s))
+        } else {
+            self.default_criteria().clone()
+        }
+    }
+    fn criteria_from_list<'b>(&self, list: impl IntoIterator<Item = &'b str>) -> CriteriaSet {
+        let mut result = self.no_criteria();
+        for criteria in list {
+            result.set_criteria(self.index[criteria])
+        }
+        result
+    }
+    fn set_criteria(&self, set: &mut CriteriaSet, criteria: &str) {
+        set.set_criteria(self.index[criteria])
+    }
+
+    fn _criteria<'b>(
+        &'b self,
+        set: &'b CriteriaSet,
+    ) -> impl Iterator<Item = (&'a str, &'a CriteriaEntry)> + 'b {
+        self.list
+            .iter()
+            .copied()
+            .enumerate()
+            .filter_map(|(idx, payload)| {
+                if set._has_criteria(idx) {
+                    Some(payload)
+                } else {
+                    None
+                }
+            })
+    }
+    fn len(&self) -> usize {
+        self.list.len()
+    }
+    fn default_criteria(&self) -> &CriteriaSet {
+        &self.default_criteria
+    }
+    fn no_criteria(&self) -> CriteriaSet {
+        CriteriaSet::none(self.len())
+    }
+    fn all_criteria(&self) -> CriteriaSet {
+        CriteriaSet::all(self.len())
+    }
+}
+
+impl CriteriaSet {
+    fn none(count: usize) -> Self {
+        assert!(
+            count <= MAX_CRITERIA,
+            "{MAX_CRITERIA} was not Enough For Everyone ({count} criteria)"
+        );
+        CriteriaSet(0)
+    }
+    fn all(count: usize) -> Self {
+        assert!(
+            count <= MAX_CRITERIA,
+            "{MAX_CRITERIA} was not Enough For Everyone ({count} criteria)"
+        );
+        if count == MAX_CRITERIA {
+            CriteriaSet(!0)
+        } else {
+            // Bit Magic to get 'count' 1's
+            CriteriaSet((1 << count) - 1)
+        }
+    }
+    fn set_criteria(&mut self, idx: usize) {
+        self.0 |= 1 << idx;
+    }
+    fn _has_criteria(&self, idx: usize) -> bool {
+        (self.0 & (1 << idx)) != 0
+    }
+    fn intersected_with(&mut self, other: &CriteriaSet) {
+        self.0 &= other.0;
+    }
+    fn unioned_with(&mut self, other: &CriteriaSet) {
+        self.0 |= other.0;
+    }
+    fn contains(&self, other: &CriteriaSet) -> bool {
+        (self.0 & other.0) == other.0
+    }
+    fn is_empty(&self) -> bool {
+        self.0 == 0
+    }
+}
+
+fn cmd_vet(out: &mut dyn Write, cfg: &Config) -> Result<(), VetError> {
+    // Not sure which we want, so make it configurable to test.
+    // Determines whether a delta must be == unaudited or just <=
+    let unaudited_matching_is_strict = true;
```

Comment:
Intention is for it to be strict.

Reply:
I can't remember where I wrote this down but I'm not convinced this is the right choice. If you say "unaudited=v3", you're basically saying "i have punted on this up to v3". So if someone comes along with an audit from v2->v5, why *would* you reject that as a proof of v5? v3 isn't at all special to you, right? 

We just want it to be a strict version number to force you to audit whenever you update the dep with no proof at all, right? The main thing you need to ensure is strict equality for "delta-less" acceptances (i.e. for when cur_version == unaudited).

Reply:
Consider the scenario where v1 had a vulnerability, which was fixed in v2, and you're using v3 (via unaudited). An attacker could theoretically reintroduce the vulnerable code in v4. In this situation, the v1->v4 diff would be clean, but the v3->v4 diff would introduce a vulnerability.

I don't claim that this scenario is particularly likely, but I think `unaudited_matching_is_strict=false` is a sort of surprising optimization that's not 100% sound.

Reply:
Intuitively, I think having something in unaudited means "I am going to take the risk of assuming this version is safe", not "I will assume this and all prior versions are safe".

Reply:
alright, fair enough. I think it's still fine to leave in here to mess around with